packages/victory-core/src/victory-state/victory-provider.tsx

```diff
@@ -0,0 +1,96 @@
+import * as React from "react";
+import { createContext, useContextSelector } from "use-context-selector";
+import { D3ScaleFn, DomainTuple, ForAxes } from "../types/prop-types";
+import { VictoryProviderProps } from "./types";
+import { FormattedDatum, getData } from "./helpers/get-data";
+import { getDomain } from "./helpers/get-domain";
+import { getRange } from "./helpers/get-range";
+import { getScale } from "./helpers/get-scale";
+
+type ScaleType = Required<ForAxes<D3ScaleFn>>;
+type DomainType = Required<ForAxes<DomainTuple>>;
+
+export interface ContextType {
+  data: FormattedDatum[];
+  scale: ScaleType;
+  domain: DomainType;
+}
+
+const VictoryContext = createContext<ContextType | null>(null);
+
+export function VictoryProvider({
+  children,
+  includeZero,
+  ...props
+}: VictoryProviderProps) {
+  // TODO: Get data
+  const domain = React.useMemo(
+    () => ({
+      x: getDomain(props, "x", includeZero),
+      y: getDomain(props, "y", includeZero)
+    }),
+    [props, includeZero]
+  );
+
+  const range = React.useMemo(
+    () => ({
+      x: getRange(props, "x"),
+      y: getRange(props, "y")
+    }),
+    [props]
+  );
+
+  const scale = React.useMemo(() => {
+    const xBaseScaleFn = getScale(props, "x");
+    const yBaseScaleFn = getScale(props, "y");
+
+    // @ts-expect-error: This is a valid scale function
+    const xScaleFn = xBaseScaleFn().domain(domain.x).range(range.x);
+    // @ts-expect-error: This is a valid scale function
+    const yScaleFn = yBaseScaleFn().domain(domain.y).range(range.y);
+
+    return {
+      x: xScaleFn,
+      y: yScaleFn
+    };
+  }, [props, domain, range]);
+
+  const data = React.useMemo(() => getData(props), [props]);
+
+  const value = {
+    scale,
+    data,
+    domain
+  };
+
+  return (
+    <VictoryContext.Provider value={value}>{children}</VictoryContext.Provider>
+  );
+}
+
+type ContextValue = ContextType | null;
+
+export function useVictoryContext<T>(selector: (value: ContextValue) => T): T {
+  const context = useContextSelector<ContextValue, ContextValue>(
+    VictoryContext,
+    (v) => v
+  );
+
+  if (!context) {
+    throw new Error("useVictoryContext must be used within a VictoryProvider");
+  }
+
+  return useContextSelector<ContextValue, T>(VictoryContext, selector);
+}
+
+export function useScale() {
+  return useVictoryContext<ScaleType>((value) => value!.scale);
```

Comment:
It's a little annoying that we're null-checking the context value, and also asserting that it is non-null when using the value. There should be a better way to type this...

Reply:
It's good enough for now.  Maybe this would do it?
```suggestion
  return useVictoryContext<ScaleType>((value) => value!.scale!);
```

Reply:
I think just the null check on the `value` is sufficient. The reason I kept the default value of the context `null` rather than something like `Symbol('no context')` is because I would have needed to check the type of the context every time, and at least a non-null assertion like `!` is easier than a typeof conditional in every single selector.

Reply:
Nevermind, I think I figured it out. I just needed to make sure the value is typed to be non-nullable in the selector type signature.